docking/pose_dock.py

```diff
@@ -1,57 +1,67 @@
 import os
 import numpy as np
 
-from rdkit import Chem
-import simtk.openmm
 from simtk.openmm import app
 from simtk.openmm.app import PDBFile
-from fe.pdb_writer import PDBWriter
+from rdkit import Chem
+from rdkit.Chem.rdmolfiles import PDBWriter, SDWriter
+from rdkit.Geometry import Point3D
 
+from fe.utils import to_md_units
 from docking import dock_setup
 from ff.handlers.deserialize import deserialize_handlers
 from timemachine.lib import LangevinIntegrator
 from timemachine.lib import custom_ops
-from io import StringIO
-
-from fe import system
-from fe.utils import to_md_units
-
-from matplotlib import pyplot as plt
 
 
 def pose_dock(
     guests_sdfile,
     host_pdbfile,
-    outdir,
+    transition_type,
     n_steps,
-    lowering_steps,
-    start_lambda,
+    transition_steps,
+    max_lambda,
+    outdir,
     random_rotation=False,
     constant_atoms=[],
+    skip_errors=False,
 ):
-    """Poses guests into a host by running short simulations in which the guests phase in over time
+    """Runs short simulations in which the guests phase in or out over time
 
     Parameters
     ----------
 
     guests_sdfile: path to input sdf with guests to pose/dock
     host_pdbfile: path to host pdb file to dock into
+    transition_type: "insertion" or "deletion"
+    n_steps: how many total steps of simulation to do (recommended: <= 1000)
+    transition_steps: how many steps to insert/delete the guest over (recommended: <= 500)
+        (must be <= n_steps)
+    max_lambda: lambda value the guest should insert from or delete to
+        (recommended: 1.1) (must be >1 for work calculation to be applicable)
     outdir: where to write output (will be created if it does not already exist)
-    n_steps: how many total steps of simulation to do (recommended: <= 20000)
-    lowering_steps: how many steps to lower the guest over (recommended: <= 10000)
-        (should be <= n_steps)
-    start_lambda: what lambda value the guest should start out at (recommended: 0.25)
-    random_rotation: whether to apply a random rotation to each guest before beginning the simulation
+    random_rotation: whether to apply a random rotation to each guest before inserting
     constant_atoms: atom numbers from the host_pdbfile to hold mostly fixed across the simulation
         (1-indexed, like PDB files)
+    skip_errors: if True, will report errors to stdout and continue on to the next guest.
+        If False, will halt upon errors.
 
     Output
     ------
 
-    A pdb file every 1000 steps (outdir/<guest_name>_<step>.pdb)
-    stdout every 1000 steps noting the step number, lambda value, and energy
+    A pdb file every 100 steps (outdir/<guest_name>_<step>.pdb)
+    stdout every 100 steps noting the step number, lambda value, and energy
+    stdout for each guest noting the work of transition
+
+    Note
+    ----
+    If the energy becomes positive, the simulation for that guest will stop and the work will not be
```

Comment:
I'm not sure if positive energy is the best way to detect im(ex)ploded systems

Reply:
the total energy is a usually system-dependent number - better way would be to check that the norm of force per atom is no more than say 10000kJ/(mol*nm)

Reply:
done